PR: https://github.com/odoo/odoo/pull/77065

From: 17e58758de7b6b02216336111d90665006a41234
From: Kartik Chavda
Date: 2021-11-26 13:37:58

Structural Changes: 3
Total Changes: 9

[IMP] sale,sale_(project,timesheet): improve the UI

Purpose of the commit is to do the generic improvement for the
project:

So in this commit, done the following changes:
- duplicating a project, copy all milestone of project and
set is_reached and deadline field as 'False'.
- chanage the order of the state button in sale order form view.
- pass mail body as default description value in message_new
method of project task.

task-2652943

closes odoo/odoo#77065

Related: odoo/enterprise#21199
Signed-off-by: Laurent Stukkens (ltu) <ltu@odoo.com>

================================= pseudo patch: =================================

--- a/addons/project/models/project.py
+++ b/addons/project/models/project.py
@@ -308,7 +308,7 @@ class Project(models.Model):
         ('to_define', 'Set Status'),
     ], default='to_define', compute='_compute_last_update_status', store=True, readonly=False, required=True)
     last_update_color = fields.Integer(compute='_compute_last_update_color')
-    milestone_ids = fields.One2many('project.milestone', 'project_id')
+    milestone_ids = fields.One2many('project.milestone', 'project_id', copy=True)
     milestone_count = fields.Integer(compute='_compute_milestone_count')
 
     _sql_constraints = [
@@ -1994,7 +1994,8 @@ class Task(models.Model):
             'name': msg.get('subject') or _("No Subject"),
             'email_from': msg.get('from'),
             'planned_hours': 0.0,
-            'partner_id': msg.get('author_id')
+            'partner_id': msg.get('author_id'),
+            'description': msg.get('body'),
         }
         defaults.update(custom_values)
 

--- a/addons/project/models/project_milestone.py
+++ b/addons/project/models/project_milestone.py
@@ -14,8 +14,8 @@ class ProjectMilestone(models.Model):
 
     name = fields.Char(required=True)
     project_id = fields.Many2one('project.project', required=True, default=_get_default_project_id)
-    deadline = fields.Date(tracking=True)
-    is_reached = fields.Boolean(string="Reached", default=False)
+    deadline = fields.Date(tracking=True, copy=False)
+    is_reached = fields.Boolean(string="Reached", default=False, copy=False)
     reached_date = fields.Date(compute='_compute_reached_date', store=True)
 
     # computed non-stored fields
